src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2047,7 +2047,376 @@ cdef class Polynomial(CommutativePolynomial):
+            (x^2 + (z8^5 + z8^3 + z8^2 + 1)*x + z8^4 + z8^3 + z8 + 1, 2),
+            (x^3 + z8^7 + z8^4 + z8^3 + z8^2 + z8, 3)]
+        """
+        q = self.base_ring().order() # p^k
+        R, x = self.parent().objgen()
+
+        # Initialise values
+        v = self
+        w = x
+        d = 0
+        e = v.degree()
+
+        # Iterate over all possible degrees with degree
+        while 2*(d + 1) <= e:
+            d = d + 1
+            w = pow(w, q, v)
+
+            ad = v.gcd(w - x)
+            yield (ad, d)
+
+            if not ad.is_one():
+                v = v // ad
+                w = w % v
+
+            e = v.degree()
+
+        # Last case, v itself might be irreducible
+        if e > 0:
+            yield (v, e)
+        return
+
+    def _cantor_zassenhaus_split_to_irreducible(self, degree):
+        """
+        Helper function for any_irreducible_factor which computes
+        a factor from a polynomial of the form `self = prod g_i(x)`
+        with all `g_i(x)` having the same degree. Uses the Cantor
+        Zassenhaus splitting method.
+
+        EXAMPLES::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F = GF(163)
+            sage: R.<x> = F[]
+            sage: f = (x^2 + 36*x + 34) * (x^2 + 36*x + 132) * (x^2 + 112*x + 48)
+            sage: f._cantor_zassenhaus_split_to_irreducible(2)   # random
+            x^2 + 112*x + 48
+            sage: f._cantor_zassenhaus_split_to_irreducible(2)   # random
+            x^2 + 36*x + 34
+            sage: f._cantor_zassenhaus_split_to_irreducible(2)   # random
+            x^2 + 36*x + 132
+            sage: factor = f._cantor_zassenhaus_split_to_irreducible(2)
+            sage: factor.is_irreducible()
+            True
+            sage: f % factor == 0
+            True
+
+        ::
+
+            sage: # needs sage.rings.finite_rings
+            sage: F.<z4> = GF(2^4)
+            sage: R.<x> = F[]
+            sage: f = (x + z4^2 + z4) * (x + z4^2 + z4 + 1)
+            sage: f._cantor_zassenhaus_split_to_irreducible(1)   # random
+            x + z4^2 + z4
+            sage: f._cantor_zassenhaus_split_to_irreducible(1)   # random
+            x + z4^2 + z4
+            sage: f._cantor_zassenhaus_split_to_irreducible(1)   # random
+            x + z4^2 + z4 + 1
+            sage: factor = f._cantor_zassenhaus_split_to_irreducible(1)
+            sage: factor.is_irreducible()
+            True
+            sage: f % factor == 0
+            True
+        """
+        R = self.parent()
+        q = self.base_ring().order()
+
+        # Polynomial is already irreducible by the assumptions of the function
+        if self.degree() == degree:
+            return self
+
+        # We expect to succeed with greater than 1/2 probability,
+        #so if we try 1000 times and fail, there's a bug somewhere.
+        for _ in range(1000):
+            # Sample a polynomial uniformly from R
+            T = R.random_element(2*degree + 1).monic()
```

Comment:
This way of sampling is essentially a temporary workaround until #37118 is done. Maybe add a comment to that effect?